build-package.sh: adding compilation based on glibc

Maxython opened the pull request:
~~only works on Termux devices~~
~~work on Termux fully developed, another part of development begins - compilation on non-Termux~~
Done

Grimler91 commented on scripts/setup-cgct.sh:
```diff
@@ -0,0 +1,25 @@
+
+if [ ! -d "$CGCT_DIR" ]; then
+	curl "${REPO_URL}/x86_64/gpkg-dev.json" -o "${TMPDIR_CGCT}/cgct.json"
+	for pkgname in cbt cgt; do
```
What does cbt and cgt stand for?

Maxython replied:
cbt - Cross Binutils for Termux (https://github.com/termux-pacman/glibc-packages/blob/main/gpkg-dev/cbt/PKGBUILD)
cgt - Cross GCCs for Termux (https://github.com/termux-pacman/glibc-packages/blob/main/gpkg-dev/cgt/PKGBUILD)

Grimler91 commented on scripts/build/termux_step_setup_toolchain.sh:
```diff
@@ -1,17 +1,23 @@
+			termux_error_exit "We do not have a setup_toolchain function for NDK version $TERMUX_NDK_VERSION"
+		fi
+	elif [ "$TERMUX_PACKAGE_LIBRARY" = "glibc" ]; then
+		# What is TERMUX_STANDALONE_TOOLCHAIN???
```
It's the path we put the toolchain in, doesn't seem like it is used here

Maxython replied:
So I understand that in the case of compilation based on the glibc library, it is useless

Grimler91 commented on scripts/setup-cgct.sh:
```diff
@@ -0,0 +1,25 @@
+		if [ ! -f "${TMPDIR_CGCT}/${filename}" ]; then
+			curl "${REPO_URL}/x86_64/${filename}" -o "${TMPDIR_CGCT}/${filename}"
+		fi
+		tar xJf "${TMPDIR_CGCT}/${filename}" -C / data
```
Seems to me like the binaries in the x86_64 package are compiled to run on x86_64, but they are being downloaded independent of arch here. Don't think it will work on any other arch than x86_64.

Maxython replied:
well cgct should only work in x86_64...
or do you mean that there need to add an arch check?

Grimler91 commented on scripts/setup-cgct.sh:
```diff
@@ -0,0 +1,25 @@
+		if [ ! -f "${TMPDIR_CGCT}/${filename}" ]; then
+			curl "${REPO_URL}/x86_64/${filename}" -o "${TMPDIR_CGCT}/${filename}"
+		fi
+		tar xJf "${TMPDIR_CGCT}/${filename}" -C / data
```
Retrieving the latest version (and not a version tagged in the file) will make it harder to do reproducible builds. In all other build scripts we use a certain (tested) version.

Maxython replied:
sounds logical

Grimler91 commented on scripts/build/termux_step_setup_variables.sh:
```diff
@@ -51,15 +66,43 @@ termux_step_setup_variables() {
+			if [ -d "${CGCT_DIR}/${TERMUX_ARCH}/bin" ]; then
+				export PATH="${CGCT_DIR}/${TERMUX_ARCH}/bin:${PATH}"
+			else
+				termux_error_exit "The cgct tools were not found"
```
Would be good to tell the users how to install them. The setup-cgct script is pretty secretive as it looks now

Maxython replied:
ok

Grimler91 commented on scripts/build/termux_step_massage.sh:
```diff
@@ -97,11 +111,29 @@ termux_step_massage() {
+						root_file="$j"
+						continue
+					fi
+					ln -sf "${TERMUX_PREFIX_CLASSICAL}/${root_file:2}" "${j}"
```
There's also no control over which file will be full file, and which will be symlink. Find doesn't sort output so symlink vs full file will vary between builds

Maxython replied:
Hard links are sorted - https://github.com/termux/termux-packages/pull/16901/files#diff-16d956212adc88d9e2d11e83f8f76549851ba7df6c22552146fd376176cb969cR119

Grimler91 commented on scripts/properties.sh:
```diff
@@ -35,25 +35,35 @@ TERMUX_BASE_DIR="/data/data/${TERMUX_APP_PACKAGE}/files"
 TERMUX_APPS_DIR="${TERMUX_BASE_DIR}/apps"
-TERMUX_PREFIX="${TERMUX_BASE_DIR}/usr"
+TERMUX_PREFIX_CLASSICAL="${TERMUX_BASE_DIR}/usr"
+TERMUX_PREFIX="${TERMUX_PREFIX_CLASSICAL}"
```
I don't like this change. Can't we just append -glibc to TERMUX_PREFIX for glibc build?

Maxython replied:
> Can't we just append -glibc to TERMUX_PREFIX for glibc build?

Well, something similar happens here - https://github.com/termux/termux-packages/pull/16901/files#diff-feabe82b48e4ea7cb1e38dddf04635abd70d3d1943332adfd218788ec36ed02aR38

This change allows: to have access to the usual `TERMUX PREFIX` ([example](https://github.com/termux-pacman/glibc-packages/blob/gpkg/gpkg/glibc-runner/build.sh)) and to simplify the logic in `build-packages.sh` scripts. Roughly speaking, this change creates two types of `TERMUX_PREFIX`: static, which in theory should not be changed for the correct operation of the compiler, and changing, which changes depending on which library it will be based on for the correct operation of the package.

Maxython commented on scripts/build/termux_step_massage.sh:
```diff
@@ -97,11 +111,29 @@ termux_step_massage() {
+						root_file="$j"
+						continue
+					fi
+					ln -sf "${TERMUX_PREFIX_CLASSICAL}/${root_file:2}" "${j}"
```
> Why this approach?

Because I don't know how to properly force the compiler to use symbolic links instead of hard ones. Of course, I can use patches, but that sounds like a so-so idea.

> What happens if three (or more) files are hardlinks?

This algorithm assumes such cases, so it will work. But if you briefly explain, then the first hardlink remains as the original and all the rest become symbolic links to that original.

Maxython commented on scripts/build/termux_step_replace_guess_scripts.sh:
```diff
@@ -4,4 +4,11 @@ termux_step_replace_guess_scripts() {
+	if [ "$TERMUX_ON_DEVICE_BUILD" = "true" ] && [ "$TERMUX_PACKAGE_LIBRARY" = "glibc" ]; then
+		local list_files=$(grep -s -r -l '^#!.*/bin/' $TERMUX_PKG_SRCDIR)
+		if [ -n "$list_files" ]; then
+			sed -i "s|#\!.*/bin/|#\!$TERMUX_PREFIX_CLASSICAL/bin/|" $list_files
```
> Which glibc specific files does this fix?

This algorithm fixes scripts that could theoretically run at compile time. For understanding, at the time of compiling something based on the glibc library, the value of `LD_PRELOAD` is removed. This is necessary because the bionic-based library is specified there and because of this, a conflict occurs during the launch of glibc commands.

Maxython commented:
And so, I did everything I wanted and in fact it is ready. Here are the main changes:
- Added the ability to specify the library base of a package using the `--library` flag (takes only two values: `bionic` and `glibc`, `bionic` by default). The compiler does not change how it works if the package is based on the `bionic` library, but with the value `glibc` there are some changes:
  - The package name will be prefixed with `-glibc` at the end if there is no `glibc` prefix in any part of the package name (for example: `bash -> bash-glibc`, `glibc-runner -> glibc-runner`)
  - The value of `TERMUX_PREFIX` will be changed in `$TERMUX_PREFIX/glibc` (`/data/data/com.termux/files/usr/glibc`) to avoid conflict.
  - The `build-package.sh` compiler sets up the environment for the tools to work correctly during compilation.
- The logic for getting package dependencies has been changed. Now if the requested package name has a `glibc` prefix anywhere, then the given package will be based on the `glibc` library. If not, then based on `bionic`. Such logic will allow to have flexible configuration of package dependencies for proper compilation and for proper operation of the package. But if you want all requested packages to be based on the same library, not depending on this logic, then you can set the `-L` flag, which means the global use of one library. It is worth noting that the `-L` flag is most likely needed for experimentation or for tests in compiling packages based on `glibc`, which are usually compiled on the basis of `bionic`.
- Added the ability to specify the main package format for repos in the `repo.json` file. This allows packages to be installed in a specific format and work properly with repos links and package databases. Set using the `pkg_format` value and can only have two values, `debian` and `pacman`. If no value is specified, it will default to `debian`.

**I note one important thing:** at the moment, when installing dependencies in the pacman format, there is no version checking. This was done on purpose so that the compiler could compile the base packages correctly. When these base packages are compiled and released, package version checking will be added.

Test compiling glibc packages - https://github.com/termux-pacman/glibc-packages/pull/82

CC @sylirre @Grimler91 @agnostic-apollo
CC @licy183 (as one of the managing TUR)
